### ISSUE TITLE
Enabling plymouth hides ZFS password prompt

### ISSUE THREAD

**Issue by ElvishJerricco** [NixOS/nixpkgs#44965]
## Issue description

Enabling the `boot.plymouth.enable` option on a system with a ZFS root with encryption enabled will cause the password prompt to be hidden at boot. You can still blindly type the password in, but this is not obvious.

## Technical details

We do have [something for LUKS](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/system/boot/plymouth.nix#L139), but it seems ZFS prompts at a different time in the lifecycle.

As an aside, plymouth does have an option for prompting for a password graphically. It'd be neat if we could use this for encrypted file systems. Then we could hide *all* text output and just show the splash.

**Comment by jtojnar** [comment]
For the record, it does not show the graphical prompt with LUKS either: https://github.com/NixOS/nixpkgs/issues/26722

**Comment by Infinisil** [comment]
I just ran into this when I enabled encryption yesterday. I discussed potential solutions with @samueldr on IRC. The most elegant imo would be to restructure the initrd script into something similar to `activationScripts`, where you can set dependencies between phases, preferably using @rycee's [DAG](https://github.com/rycee/home-manager/blob/master/modules/lib/dag.nix) library.

**Comment by cleverca22** [comment]
2 solutions for this:
1: https://wiki.ubuntu.com/Plymouth#A.22plymouth_ask-for-password.22
2:
 ```
[root@amd-nixos:~]# plymouth --help
  show-splash        Tell daemon to show splash screen
  hide-splash        Tell daemon to hide splash screen
```
you can just hide/show the splash, to let the text console prompt things

**Comment by cole-h** [comment]
I believe https://github.com/NixOS/nixpkgs/pull/88789 will fix this once it's merged. I just rebooted my system and found that I was not using that branch, because there was no prompt and I had to blindly enter my password. However, using that branch, I was prompted for my password via plymouth as one would hope/expect.

**Comment by nh2** [comment]
Still important to me

**Comment by misuzu** [comment]
Bump.

**Comment by Majiir** [comment]
This is now working on `nixos-unstable` (after https://github.com/NixOS/nixpkgs/pull/179619 merged) as long as you also set `boot.initrd.systemd.enable = true` in your config. Fair warning, that option says:

> Note: This is in very early development and is highly experimental. Most of the features NixOS supports in initrd are not yet supported by the intrd generated with this option.

**Comment by ElvishJerricco** [comment]
That note as it's currently written is *a little* outdated. It's definitely still experimental, but I wouldn't say *highly* experimental anymore. Quite a few of us actually daily drive it with a variety of fancy initrd use cases. And at this point we do actually support most of what the old scripted initrd supports; we're pretty much just missing networking (#169116 is near completion for that), and some of the fancier LUKS options.

**Comment by Majiir** [comment]
Closing this as completed. For future readers, ZFS+Plymouth works if you set `boot.initrd.systemd.enable = true`. This enables the systemd initrd, which replaces the older script-based stage 1 initrd.

See https://github.com/NixOS/nixpkgs/issues/26722 for a similar issue about LUKS+Plymouth.

**Comment by ElvishJerricco** [comment]
@Majiir I disagree with closing this. Systemd initrd is still marked experimental and is not feature complete. The default initrd in NixOS has this issue.
